### ISSUE TITLE
Policy out softmax with illegal moves

### ISSUE THREAD

**Issue by apollo-time** [mokemokechicken/reversi-alpha-zero#22]
https://github.com/mokemokechicken/reversi-alpha-zero/blob/5ee2f330663b34513f0c894eb658f03a1201f400/src/reversi_zero/agent/model.py#L48

I see calculate policy softmax on the all moves contains illegal.
How can calculate softmax on the only legal moves, if set placeholder for legal moves?

**Comment by mokemokechicken** [comment]
> I see calculate policy softmax on the all moves contains illegal.
> How can calculate softmax on the only legal moves, if set placeholder for legal moves?

Does this answer understand the intent of the question?

----

For example, 

```python
  # no output for 'pass'
  policy_out = Dense(8*8, kernel_regularizer=l2(mc.l2_reg), activation="softmax", name="policy_out")(x)
```

↓

```python
  legal_mask = Input((8 * 8))  # (0: illegal, 1: legal)
  ...
  # no output for 'pass'
  x = Dense(8*8, kernel_regularizer=l2(mc.l2_reg))(x)
  x = Multiply()([x, legal_mask])
  policy_out = Activation("softmax", name="policy_out")(x)
  
  ...
  
  self.model = Model([in_x, legal_mask], [policy_out, value_out], name="reversi_model")

```

Input of `legal_mask` is required to be computed in all training data.

**Comment by apollo-time** [comment]
Is not equal softmax((0, -0.5, 0.5))[1:2] and softmax((-0.5,0.5)) when legal_mask=(0,1,1)?

**Comment by gooooloo** [comment]
@apollo-time Mathematically it is. But in neural network, you have same size of output elements no matter what input it is, that is 3 in your (0, -0.5, 0.5) case. You can only apply the legal_mask after the neural network output.

**Comment by mokemokechicken** [comment]
@apollo-time 

oh, I was careless.
How about like this?

```
  legal_mask = Input((8 * 8))  # (0: illegal, 1: legal)
  legal_mask = Lambda(lambda x: (x-1)*1000000)(legal_mask)  # illegal -> -1000000, legal -> 0
  ...
  # no output for 'pass'
  x = Dense(8*8, kernel_regularizer=l2(mc.l2_reg))(x)
  x = Add()([x, legal_mask])
  policy_out = Activation("softmax", name="policy_out")(x)
  
  ...
  
  self.model = Model([in_x, legal_mask], [policy_out, value_out], name="reversi_model")
```


```
> softmax([-0.5, 0.5])
[ 0.26894142,  0.73105858]

> softmax([-0.5, 0.5, -1000000])
[ 0.26894142,  0.73105858,  0.        ]
```

**Comment by apollo-time** [comment]
right, just it. thanks.
